### PR TITLE
feat: user agent enhancements - cross-category api>api-graphql>pubsub

### PR DESCRIPTION
**Pull request by erinleigh90**
#### Description of changes
Creates path to send customUserAgentDetails from Datastore, through API, GraphQLAPI, and PubSub. This was accomplished without changes to the customer-facing API by creating an internal API, GraphQLAPI, and PubSub class, where the methods take an additional customUserAgentDetails parameter and pass that on to the final service call.

This change makes updates to withSSRContext for DataStore, adding InternalAPI to the SSR context when DataStore is added.

I did not create an internal version of AWSAppSyncRealTimeProvider, since the expected use is internal only, although it is exported from the PubSub category.

There is an issue with large bundle size increase in GraphQL: the size increased by 60.26 kb, from 105.6 kB to 165.85 kB. investigation in progress.

#### Description of how you validated changes

This was tested using sample apps on React and NextJS for DataStore, API, and PubSub, using app-linking and validating that the correct user agent was sent with each graphql operation and websocket.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/api/src/internal/InternalAPI.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+		return this._graphqlApi.graphql(
+			options,
+			additionalHeaders,
+			apiUserAgentDetails
+		);
+	}
+}
+
+export const InternalAPI = new InternalAPIClass(null);
```
Comment (stocaaro) [inline packages/api/src/internal/InternalAPI.ts:248]: It seems significant that we now have 2 Instances for API / API GraphQL / PubSub. I can't think of any specific issues this might cause, but it seems possible that there may be issues.

Isolated instances will hold isolated state. Are there any use-cases where a customer using the public instance could have problems because the state maintained by the internal instance state isn't shared?

- Is there any way we might end up with multiple websockets for appsync because of this?
- Could having two instances with two GraphQL Cache's cause duplicate work / memory usage / access for a customer mixing datastore with normal API.graphql?

I think the answers I'm coming up with is that these are long tail concerns that probably won't be problems, but worth some manual testing effort to look for oddities before we roll out custom user agent.

**This comment is meant to act as an invitation for other folks to think through potential implications and raise any concerns.**

**Review by stocaaro — APPROVED** [review]
Overall, this change looks great. I've made a handful of comments, but nothing that would gate getting this merged.  Interested to get input from a data engineer or two and getting input from Francisco who has a lot of time with this integration space and then maybe Allan re: the submodule surfacing work.

File: packages/pubsub/src/internal/InternalPubSub.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+// import '../Common/Polyfills';
```
[the remark this answers is not in the document]
Comment (erinleigh90) [reply packages/pubsub/src/internal/InternalPubSub.ts:3]: Yes

File: packages/pubsub/src/Providers/AWSAppSyncRealTimeProvider/index.ts [hunk 3]
```diff
@@ -208,7 +210,8 @@ export class AWSAppSyncRealTimeProvider extends AbstractPubSubProvider<AWSAppSyn
 
 	subscribe(
 		_topics: string[] | string,
-		options?: AWSAppSyncRealTimeProviderOptions
+		options?: AWSAppSyncRealTimeProviderOptions,
+		customUserAgentDetails?: CustomUserAgentDetails
```
[the remark this answers is not in the document]
Comment (erinleigh90) [reply packages/pubsub/src/Providers/AWSAppSyncRealTimeProvider/index.ts:214]: Yes, would appreciate some weigh-in on this conversation if possible. It doesn't seem worthwhile to create an internal version of a basically internal class, but that opinion may not be shared.

File: packages/api/src/internal/InternalAPI.ts (hunk 1 above)
Comment (erinleigh90) [reply packages/api/src/internal/InternalAPI.ts:248]: Absolutely

**Review by svidgen — APPROVED** [review]

**Review by svidgen — APPROVED** [review]

File: packages/api-graphql/internal/package.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+	"name": "@aws-amplify/api-graphql/internal",
+	"types": "../lib-esm/internal/index.d.ts",
+	"main": "../lib/internal/index.js",
+	"module": "../lib-esm/intternal/index.js",
```
Comment (AllanZhengYP) [inline packages/api-graphql/internal/package.json:5]: Typo: -> `internal`
Also maybe we should use plurals

Comment (AllanZhengYP) [reply packages/api-graphql/internal/package.json:5]: `"react-native"` entry is optional here, but better to make sure it works on react-native, or add react-native in every internal path.

File: packages/core/src/Platform/types.ts [hunk 5]
```diff
@@ -31,8 +31,8 @@ export enum Category {
 	Analytics = 'analytics',
 	DataStore = 'datastore',
 	Geo = 'geo',
-	Interactions = 'interactions',
 	InAppMessaging = 'inappmessaging',
+	Interactions = 'interactions',
```
Comment (AllanZhengYP) [inline packages/core/src/Platform/types.ts:35]: Super nit: unrelated change?

File: packages/pubsub/internal/package.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
+	"name": "@aws-amplify/pubsub/internal",
+	"types": "../lib-esm/internal/index.d.ts",
+	"main": "../lib/internal/index.js",
+	"module": "../lib-esm/intternal/index.js",
```
Comment (AllanZhengYP) [inline packages/pubsub/internal/package.json:5]: Ditto

**Review by AllanZhengYP — COMMENTED** [review]
I'm halfway through it but left some comments around the internal subpath, so you can address them async.

File: packages/datastore/src/sync/processors/mutation.ts [hunk 7]
```diff
@@ -344,10 +349,19 @@ class MutationProcessor {
 
 				const opType = this.opTypeFromTransformerOperation(operation);
 
+				const customUserAgentDetails: CustomUserAgentDetails = {
+					category: Category.DataStore,
+					action: DataStoreAction.GraphQL,
+				};
```
[the remark this answers is not in the document]
Comment (erinleigh90) [reply packages/datastore/src/sync/processors/mutation.ts:355]: We want to be able to manage the action values for each category independently so overlapping action values are clearly differentiated in the code based on category. That way seeing the Auth action "SignIn" with a value of "1" is not confused with the API action GraphQL with a value of "1"

That being said, DataStore is a bit tricky in terms of action, since a customer action doesn't directly initiate the service call, so we compromised in datastore and separated the "Subscribe" action from calls made directly via API.graphql. We want to manage everything the same way in every category as much as possible, so while DataStore and API have a bit of redundency, in other categories this is not the case. Setting Datastore category to an API action seems strange to me if we are managing actions separately in every other category.

File: packages/api-graphql/src/GraphQLAPI.ts [hunk 8]
```diff
@@ -46,169 +16,11 @@ export const graphqlOperation = (
...
-	/**
-	 * Initialize GraphQL API with AWS configuration
-	 * @param {Object} options - Configuration object for API
-	 */
-	constructor(options) {
-		this._options = options;
-		logger.debug('API Options', this._options);
-	}
-
+export class GraphQLAPIClass extends InternalGraphQLAPIClass {
```
[the remark this answers is not in the document]
Comment (erinleigh90) [reply packages/api-graphql/src/GraphQLAPI.ts:19]: This was the easiest way I found of being able to have all the same methods, but override the graphql method with fewer parameters. Open to suggestions if there is a better way to accomplish this.

File: packages/core/src/Platform/types.ts (hunk 5 above)
Comment (erinleigh90) [reply packages/core/src/Platform/types.ts:35]: True, belongs to a different category, I will remove from this branch.

File: packages/core/src/Platform/types.ts [hunk 9]
```diff
@@ -45,28 +45,29 @@ export enum AnalyticsAction {
...
-export enum InAppMessagingAction {
+export enum InteractionsAction {
 	None = '0',
 }
 export enum PredictionsAction {
 	None = '0',
 }
 export enum PubSubAction {
-	None = '0',
+	Subscribe = '1',
```
[the remark this answers is not in the document]
Comment (erinleigh90) [reply packages/core/src/Platform/types.ts:70]: Discussing with Abdallah his preference

File: packages/datastore/src/sync/processors/mutation.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/datastore/src/sync/processors/mutation.ts:355]: > so overlapping action values are clearly differentiated in the code based on category

Right. I don't understand why it's not just this:

```ts
const userAgentAddendum = UserAgents.DataStore.GraphQL;
```

This keeps the gory details of how the header is constructed hidden and *far away* from the category's control, and it naturally encompasses and enforces the hierarchical nature of category/action behind "a thing that knows what all the valid user agent category/action combinations are" and so on.

I'm sure the types will keep me in line. And, I don't think I'll necessarily lose sleep over the Java-y redundancy here -- not just in seeing *DataStore* category and DataStore action, but also the `category: Category...` and `action: CategoryAction.*`. But, unless the invoking code gains something from specifying these two values independently, why not favor brevity. 😅 

This is an internal two-way door. So, you have my approval. But, the feedback is that I'm being given two dials that will yet at me if they're not set to compatible settings, whereas we could just as easily make a *single* dial that has consists entirely of valid options.

[the remark this answers is not in the document]
Comment (erinleigh90) [reply packages/datastore/src/sync/processors/mutation.ts:355]: Got it, that makes sense

File: packages/api-graphql/src/GraphQLAPI.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/api-graphql/src/GraphQLAPI.ts:19]: The alternative would just be composition. (`GraphQLAPIClass` methods creates and wraps an `InternalGraphQLAPIClass` instance.)  But, I can see your reasoning for going this route at this stage.

Thanks!

File: packages/datastore/src/sync/processors/mutation.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (stocaaro) [reply packages/datastore/src/sync/processors/mutation.ts:355]: This would make a good refactor. Been carrying this pattern forward with other categories, but since actions are inherently coupled to category, no need to label category separately.
